### PR TITLE
stm32/adc.c Provide ADC.read_timed_with() method.

### PR DESCRIPTION
**Pull request by peterhinch**
The current implementation makes it difficult to extract timing/phase information from multiple ADC channels if the sample frequency is high enough to require a timed read.

This patch provides a means of performing a timed read on an arbitrary number of ADC channels. On a timer tick each assigned channel is read into its own buffer. This occurs in quick succession before awaiting the next tick. Like **timed_read** it is a blocking method.

There are a number of use cases in the audio frequency range including echo location, determining the position of an external sound source and electronic test and measurement. Seismic applications are another possible use where readings from a pair of geophones might provide vector results.

Docs are updated. The test program is not changed pending a resolution of https://github.com/micropython/micropython/issues/3655.

### CODE REVIEW

**Comment by peterhinch** [comment]
That's a good idea. I'll try going that route.

I've done a bit of work characterising its performance with (to me) surprising results. I set a DAC up to output a continuous full scale sinewave, connecting it to one ADC directly and another either directly or via a single pole CR network to give a 45° phase lag at 20KHz (measured on a scope).

My code reads 128 12-bit samples from the two ADC's at 2*Nyquist (so I acquire 32 cycles regardless of frequency). With a 20KHz output it is therefore sampling at 80KHz. It calculates the relative phase using in-phase and quadrature measurements to provide a full 2π radians. An empirically-determined 3.236μs of phase compensation is applied - this is evidently the mean delay between the two ADC reads.

The outcome is impressively consistent between runs and over a Pyboard V1.0 and V1.1. At 20KHz the phase error relative to the scope reading varied from -1° to -3°. The worst-case error seemed to occur around 4KHz and was around 9°. At low frequencies as you'd expect the error tended to zero.

Evidently the phase detection algorithm does a good job of averaging the timing uncertainties I identified in the docs. I think this opens up interesting new application areas for the Pyboard. It's way ahead of what my network analyser achieves in Python: I stop even attempting phase measurements above 5KHz.

Of course this is the easy case where the phase detection is performed on an incoming frequency which is precisely known, so an integer number of input cycles are read. This corresponds to a subset of applications. Accommodating unknown frequencies is a matter of doing some more maths on the sample sets - I've not yet thought this through but I think a window function would be required.

**Comment by peterhinch** [comment]
I now have a method `read_timed_multi` with the new arg pattern, but I can't see how to declare it as a classmethod or staticmethod. At present I access it as an instance method with the first arg (self) ignored.

Please could you advise or point me to an existing example.

It turns out that the high performance referred to above is genuine: my method of overrun detection was overly pessimistic. Sampling two adcs at 160Ksps each is workable without overruns.

**Comment by dpgeorge** [comment]
> Please could you advise or point me to an existing example.

See for example https://github.com/micropython/micropython/blob/master/extmod/vfs_fat.c#L122

**Comment by peterhinch** [comment]
This is now updated with ``read_timed_multi`` as a static method. The test program is updated and bugs fixed. Amongst updates to ``adc.rst`` I simplified the code example for reading temperature etc. The old example is superseded given that an ADC may now be instantiated with a specific channel.

**Comment by chuckbook** [comment]
While updating the pyb.ADC.rst would it make sense to add info about read_vref() and associated corrections done with read_core_*() methods?

**Comment by peterhinch** [comment]
@chuckbook Please could you clarify what you're after: I thought I'd done this [here](https://github.com/peterhinch/micropython/blob/adc-phase/docs/library/pyb.ADC.rst).

@dpgeorge With this patch I can perform phase measurements between two channels up to 36KHz (144KHz sampling frequency). Accuracy (with 3.236μs compensation) over the range 0-34KHz is about +1.6° -3.8°.

**Comment by chuckbook** [comment]
@peterhinch All STM32 MCUs (at least those supported by ports/stm32) provide factory calibration for
both temperature and internal Vref. If MICROPY_PY_BUILTINS_FLOAT is set, adc.c already implements necessary corrections for Vref, Vbat and temperature. Also external Vref is calculated and accessible via read_vref().

**Comment by peterhinch** [comment]
The `read_vref()` method is of the `ADCAll` object. I agree this is not documented, although `read_core_vref()` is. The `ADCAll` object is problematic for the reason stated in the docs, but I can add this if you wish.

As far as I can see, without recourse to  `ADCAll`, the only way to access the temperature and voltage readings is by using `ADC` methods on the relevant channels. The class and instance methods of `ADC` (with this patch) are:
```
MicroPython v1.9.3-362-g2ad555b-dirty on 2018-03-17; PYBv1.1 with STM32F405RG
Type "help()" for more information.
>>> from pyb import ADC
>>> dir(ADC)
['__class__', '__name__', 'read', 'read_timed', 'read_timed_multi']
>>> adc = ADC(1)
>>> dir(adc)
['__class__', 'read', 'read_timed', 'read_timed_multi']
>>> 
```
It seems that you have knowledge about this that I lack. Perhaps it would save time if you submit your own PR?

**Comment by chuckbook** [comment]
I absolutely agree that the preferred way to use the `ADC` subsystem is not necessarily the `ADCAll` class.
However, for the casual user of MCU ADC it is sometimes more convenient to use the chewed results from `ADCAll`. The most annoying side effect of turning all affected gpios into analog inputs can be suppressed by giving an extra (undocumented) mask parameter when instantiating `ADCAll(res, 0x70000)`.

**Comment by peterhinch** [comment]
You've evidently studied this in depth. It would be great if you could submit a PR to document the mask parameter and to explain the benefits of using a masked `ADCAll` compared to reading individual ADC channels. There seems little merit in my duplicating research you've already done.

**Comment by chuckbook** [comment]
I created a PR (3676) which adds some info about ADCAll features.

**Comment by dpgeorge** [comment]
@peterhinch thanks for fixing this up.  I've now tested it and it works great!

I have one suggestion to slightly modify the API: I think it would be better to return an integer being the number of times the timer already triggered (ie the number of miss-timed samples), instead of just a True/False if there were any/no misses.  Then you could tell if you are just on the edge of being too fast.  Eg with 3 ADCs I could get up to around 70kHz with no misses, then at 80kHz I had 1 miss, then at 100kHz I had 6 misses, then at 110kHz I had a full 99 misses (for a buffer of length 100).

What do you think of this idea?

**Comment by dpgeorge** [comment]
BTW: if you do happen to rebase this against master you'll need to change MCU_SERIES_F4 to STM32F4 (and similar with the others).

**Comment by peterhinch** [comment]
I implemented the boolean result because we get it virtually"for free" in terms of performance. I envisage its use in development to measure the highest rate which can be used without risking a loss of phase accuracy. There is no need to check it at runtime as the code is deterministic (I ran high speed tests with interrupts disabled).

Replacing the boolean with a count, the only values which would be of interest to someone wanting to measure phase would be 0 (good) or 1 (possible loss of accuracy). A value > 1 implies a loss of entire samples which implies a complete failure rather than a loss of accuracy.

As far as I can see implementing a count would slow down the code. After each ADC reading you'd need to check whether the timer had timed out and, if so, update the count and reset the timer. The essence of this patch is to read each ADC in succession as quickly as possible to facilitate accurate measurement of relative timing/phase. So it checks whether the timer has timed out just once, at the end of a set of readings.

I'd be loath to compromise performance for a benefit which is arguably small.

File: ports/stm32/adc.c [hunk 1]
```diff
@@ -448,9 +448,145 @@ STATIC mp_obj_t adc_read_timed(mp_obj_t self_in, mp_obj_t buf_in, mp_obj_t freq_
...
+        }
+    __HAL_TIM_CLEAR_FLAG(tim, TIM_FLAG_UPDATE);
+
+    // Overrun check: assume success.
+    uint success = 1;
+    uint nelems = bufinfo.len / typesize;
+    for (uint elem_index = 0; elem_index < nelems; elem_index++) {
+        if (__HAL_TIM_GET_FLAG(tim, TIM_FLAG_UPDATE) != RESET) {
+            // Timer has already triggered.
+            success = 0;
```
Comment (dpgeorge) [inline ports/stm32/adc.c]: I was thinking to just change this line to `++num_late;`

File: ports/stm32/adc.c [hunk 2]
```diff
@@ -448,9 +448,145 @@ STATIC mp_obj_t adc_read_timed(mp_obj_t self_in, mp_obj_t buf_in, mp_obj_t freq_
...
+    uint value = ADCx->DR;
+
+    // Ensure first sample is on a timer tick
+    __HAL_TIM_CLEAR_FLAG(tim, TIM_FLAG_UPDATE);
+    while (__HAL_TIM_GET_FLAG(tim, TIM_FLAG_UPDATE) == RESET) {
+        }
+    __HAL_TIM_CLEAR_FLAG(tim, TIM_FLAG_UPDATE);
+
+    // Overrun check: assume success.
+    uint success = 1;
```
Comment (dpgeorge) [inline ports/stm32/adc.c]: ... change this line to `uint num_late = 0;`

File: ports/stm32/adc.c [hunk 3]
```diff
@@ -448,9 +448,145 @@ STATIC mp_obj_t adc_read_timed(mp_obj_t self_in, mp_obj_t buf_in, mp_obj_t freq_
...
+            mp_buffer_info_t bufinfo_curr;  // Get buf for current ADC
+            mp_get_buffer_raise(buf_array[array_index], &bufinfo_curr, MP_BUFFER_WRITE);
+            mp_binary_set_val_array_from_int(bufinfo_curr.typecode, bufinfo_curr.buf, elem_index, value);
+        }
+    }
+
+    // turn the ADC off
+    adc0 = adc_array[0];
+    HAL_ADC_Stop(&adc0->handle);
+    return mp_obj_new_bool(success);
```
Comment (dpgeorge) [inline ports/stm32/adc.c]: ... and change this line to `return MP_OBJ_NEW_SMALL_INT(num_late);`

**Comment by dpgeorge** [comment]
> A value > 1 implies a loss of entire samples which implies a complete failure rather than a loss of accuracy.

See the description of my idea inline in the PR code.  A value > 1 just indicates how many samples were sampled *late*.  No matter how late or how many are late, you will still sample `nelems` values (same with how the code currently is: even if it returns `False` it will still have made `nelems` ADC samples (for each ADC object) and stored them into the buffers).

But I guess what you are saying is that they are either all on time, or all late, that there can't really be an inbetween because that would imply that the loop has varying time over subsequent loops.  This would be true if interrupts are enabled.

> As far as I can see implementing a count would slow down the code.

See my comments inline.  The counter version should be almost identical in code size and execution time.

**Comment by peterhinch** [comment]
Ah, I thought you were proposing checkig between each sample. So this would be cost free. But
>But I guess what you are saying is that they are either all on time, or all late, that there can't really be an in between because that would imply that the loop has varying time over subsequent loops. This would be true if interrupts are enabled.

Quite. And, as I suggest in the docs, if you're doing high speed measurements it's best to disable interrupts.

I'm happy to do this, but I can't see any practical benefit.

**Comment by dpgeorge** [comment]
Ok, let's keep things as-is then, returning False/True.  You can anyway time how long the function takes (eg time.ticks_us) to see how close you are to it failing.

**Comment by dpgeorge** [comment]
Ok, this was rebased, cleaned up a little and merged in 4f40fa5cf4de7a8eabe073493e2df54c8a08ea89

I separated out the improvements to the adc test in 0096a4bd005606489a5b7fdfda8e2a60a1709e13 because they were independent to the addition of ADC.read_timed_multi().

During the rebase I wasn't able to cleanly merge the changes to the docs for the ADCAll object, which are anyway not related to this PR.  And these changes conflicted due to the recent merge of 8f11d0b532bd0203c875ad9e3e4efd8a7014ca15 .  @peterhinch would you mind submitting a PR to fix the ADCAll docs (basically remove the adcread() function)?

**Comment by dpgeorge** [comment]
In aebd9701a78d267cb264d400804b02c5b7a00e9a I further optimised ADC.read_timed_multi() by caching the buffer pointers.  This improves the maximum data rate capture by about 50%.

**Pull request closed without merging**